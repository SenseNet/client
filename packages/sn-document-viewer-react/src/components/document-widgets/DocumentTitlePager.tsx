import React, { useCallback, useState } from 'react'
import ClickAwayListener from '@material-ui/core/ClickAwayListener'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import { useDocumentData, useLocalization, useViewerState } from '../../hooks'

/**
 * Document widget component for paging
 */
export const DocumentTitlePager: React.FC = () => {
  const [isFocused, setIsFocused] = useState(false)

  const { documentData } = useDocumentData()
  const viewerState = useViewerState()
  const localization = useLocalization()

  const setPage = useCallback(
    (index: number) => {
      viewerState.updateState({ activePages: [index] })
      viewerState.onPageChange.setValue(index)
    },
    [viewerState],
  )

  const [currentPage, setCurrentPage] = useState(viewerState.activePages[0])

  const gotoPage = (page: string | number) => {
    let pageInt = typeof page === 'string' ? parseInt(page, 10) : page
    if (!isNaN(pageInt)) {
      pageInt = Math.max(pageInt, 1)
      pageInt = Math.min(pageInt, documentData.pageCount)
      setCurrentPage(pageInt)
      setPage(pageInt)
    }
  }

  return (
    <ClickAwayListener onClickAway={() => setIsFocused(false)}>
      <Typography
        onClick={() => setIsFocused(true)}
        variant="h6"
        color="inherit"
        style={{ cursor: 'pointer', display: 'flex', alignItems: 'center', overflow: 'hidden', margin: '0 2.5em' }}
        title={documentData.documentName}>
        <div style={{ whiteSpace: 'nowrap', overflow: 'hidden', textOverflow: 'ellipsis' }}>
          {documentData.documentName}&nbsp;
        </div>
        {isFocused ? (
          <form
            onSubmit={(ev) => {
              ev.preventDefault()
            }}>
            <TextField
              style={{ flexShrink: 0 }}
              title={localization.gotoPage}
              value={currentPage}
              onChange={(ev) => gotoPage(ev.currentTarget.value)}
              type="number"
              required={true}
              InputLabelProps={{
                shrink: true,
              }}
              inputProps={{
                min: 1,
                max: documentData.pageCount,
                style: { textAlign: 'center' },
              }}
              margin="dense"
            />
          </form>
        ) : (
          <div style={{ flexShrink: 0 }}>
            {currentPage} / {documentData.pageCount}
          </div>
        )}
      </Typography>
    </ClickAwayListener>
  )
}
