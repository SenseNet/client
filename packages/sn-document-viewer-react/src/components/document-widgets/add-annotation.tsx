import { createStyles, makeStyles, Theme } from '@material-ui/core'
import { ChatSharp } from '@material-ui/icons'
import clsx from 'clsx'
import React from 'react'
import { useDocumentPermissions, useLocalization, useViewerState } from '../../hooks'
import { ToggleBase } from './toggle-base'

const useStyles = makeStyles((theme: Theme) => {
  return createStyles({
    iconButton: {},
    icon: {},
    iconActive: {
      fill: theme.palette.primary.main,
    },
  })
})

type AddAnnnotationClassKey = Partial<ReturnType<typeof useStyles>>

/**
 * Document widget component to draw Annotation
 */
export const AddAnnotationWidget: React.FC<{ classes?: AddAnnnotationClassKey }> = (props) => {
  const classes = useStyles(props)
  const localization = useLocalization()
  const viewerState = useViewerState()
  const permissions = useDocumentPermissions()

  const pageRotation =
    viewerState.rotation?.find((rotation) => rotation.pageNum === viewerState.activePage)?.degree || 0

  return (
    <ToggleBase
      disabled={!permissions.canEdit || pageRotation !== 0}
      classes={classes}
      active={viewerState.activeShapePlacing === 'annotation'}
      title={localization.addAnnotation}
      setValue={(value) => viewerState.updateState({ activeShapePlacing: value ? 'annotation' : 'none' })}>
      <ChatSharp
        className={clsx(classes.icon, { [classes.iconActive]: viewerState.activeShapePlacing === 'annotation' })}
      />
    </ToggleBase>
  )
}
