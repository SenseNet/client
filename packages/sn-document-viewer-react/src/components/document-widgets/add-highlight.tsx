import { createStyles, makeStyles, Theme } from '@material-ui/core'
import ChatBubbleOutlineSharp from '@material-ui/icons/ChatBubbleOutlineSharp'
import clsx from 'clsx'
import React from 'react'
import { useDocumentPermissions, useLocalization, useViewerState } from '../../hooks'
import { ToggleBase } from './toggle-base'

const useStyles = makeStyles((theme: Theme) => {
  return createStyles({
    iconButton: {},
    icon: {},
    iconActive: {
      fill: theme.palette.primary.main,
    },
  })
})

type AddRHightlightClassKey = Partial<ReturnType<typeof useStyles>>

/**
 * Document widget component to draw Hihgligh
 */
export const AddHighlightWidget: React.FC<{ classes?: AddRHightlightClassKey }> = (props) => {
  const classes = useStyles(props)
  const localization = useLocalization()
  const viewerState = useViewerState()
  const permissions = useDocumentPermissions()

  return (
    <ToggleBase
      disabled={!permissions.canEdit}
      classes={classes}
      isVisible={viewerState.isPlacingHighlight}
      title={localization.addHighlight}
      setValue={(value) => viewerState.updateState({ isPlacingHighlight: value })}>
      <ChatBubbleOutlineSharp
        className={clsx(classes.icon, { [classes.iconActive]: viewerState.isPlacingHighlight })}
      />
    </ToggleBase>
  )
}
