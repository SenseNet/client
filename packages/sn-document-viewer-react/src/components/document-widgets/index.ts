export * from './add-annotation'
export * from './add-highlight'
export * from './add-redaction'
export * from './document-title-pager'
export * from './download-widget'
export * from './print-widget'
export * from './rotate-active-pages'
export * from './rotate-document'
export * from './rotate-widget'
export * from './save-widget'
export * from './share-widget'
export * from './toggle-base'
export * from './toggle-comments'
export * from './toggle-redaction'
export * from './toggle-shapes'
export * from './toggle-thumbnails'
export * from './toggle-watermark'
export * from './zoom-in-out'
