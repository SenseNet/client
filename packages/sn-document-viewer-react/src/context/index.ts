export * from './api-settings'
export * from './comment-states'
export * from './comments'
export * from './document-data'
export * from './document-permissions'
export * from './localization-context'
export * from './preview-image-data'
export * from './viewer-settings'
export * from './viewer-state'
