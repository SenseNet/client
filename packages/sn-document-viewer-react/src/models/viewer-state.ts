import { ObservableValue } from '@sensenet/client-utils'

/**
 * Rotation model
 */
export interface RotationModel {
  /**
   * The index of the page
   */
  pageNum: number

  /**
   * Rotation in degrees
   */
  degree: number
}

/**
 * Type model for the Viewer state
 */
export interface ViewerState {
  /**
   * The active page
   */
  activePage: number
  /**
   * Watermark is on / off
   */
  showWatermark: boolean
  /**
   * Redaction is on / off
   */
  showRedaction: boolean
  /**
   * Shapes are on / off
   */
  showShapes: boolean

  /**
   * Thumbnails on / off
   */
  showThumbnails: boolean

  /**
   * Zoom level relative to the fitted image size
   */
  zoomLevel: number

  /**
   * Rotation
   */
  rotation?: RotationModel[]

  /**
   * Determines if comments are shown
   */
  showComments: boolean

  /**
   * True if there are any changes on the document
   */
  hasChanges: boolean

  /**
   * Variable that indicates if a comment marker placing is in progress
   */
  isPlacingCommentMarker: boolean

  /**
   * Indicates if the user is creating a comment at the moment
   */
  isCreateCommentActive: boolean

  /**
   *
   */
  onPageChange: ObservableValue<number>
}
