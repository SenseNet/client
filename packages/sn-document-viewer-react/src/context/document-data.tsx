import { DocumentData } from '@sensenet/client-core'
import { deepMerge, DeepPartial, sleepAsync } from '@sensenet/client-utils'
import { useRepository } from '@sensenet/hooks-react'
import React, { useCallback, useEffect, useState } from 'react'
import Semaphore from 'semaphore-async-await'
import { POLLING_INTERVAL } from '../components'
import { PreviewState } from '../enums'
import { useDocumentViewerApi, useViewerSettings } from '../hooks'

const defaultDocumentData: DocumentData = {
  documentName: '',
  documentType: '',
  fileSizekB: 0,
  hostName: '',
  idOrPath: 0,
  pageCount: PreviewState.Loading,
  shapes: {
    annotations: [],
    highlights: [],
    redactions: [],
  },
  error: undefined,
}

export interface DocumentDataContextType {
  documentData: DocumentData
  updateDocumentData: (data: DeepPartial<DocumentData>) => void
  isInProgress: boolean
  triggerReload: () => void
}

export const defaultDocumentDataContextValue: DocumentDataContextType = {
  documentData: defaultDocumentData,
  updateDocumentData: () => undefined,
  isInProgress: false,
  triggerReload: () => {},
}

export const DocumentDataContext = React.createContext<DocumentDataContextType>(defaultDocumentDataContextValue)

export const DocumentDataProvider: React.FC = ({ children }) => {
  const api = useDocumentViewerApi()
  const doc = useViewerSettings()
  const repo = useRepository()

  const [documentData, setDocumentData] = useState<DocumentData>(defaultDocumentData)
  const [loadLock] = useState(new Semaphore(1))
  const [isInProgress, setIsInProgress] = useState(false)
  const [reloadToken, setReloadToken] = useState(0)

  const triggerReload = () => setReloadToken((prevValue) => prevValue + 1)

  useEffect(() => {
    const ac = new AbortController()
    const getData = async () => {
      const result = await api.getDocumentData({
        hostName: repo.configuration.repositoryUrl,
        idOrPath: doc.documentIdOrPath,
        version: doc.version,
        abortController: ac,
      })
      if (result.pageCount === PreviewState.Loading && ac.signal.aborted === false) {
        setDocumentData(result)
        await sleepAsync(POLLING_INTERVAL)
        getData()
      } else {
        setDocumentData(result)
      }
    }
    ;(async () => {
      try {
        setIsInProgress(true)
        await loadLock.acquire()
        await getData()
      } catch (error) {
        if (!ac.signal.aborted) {
          setDocumentData({ ...defaultDocumentData, pageCount: PreviewState.ClientFailure, error: error.toString() })
          throw error
        }
      } finally {
        setIsInProgress(false)
        loadLock.release()
      }
    })()
    return () => ac.abort()
  }, [api, doc.documentIdOrPath, doc.version, loadLock, repo.configuration.repositoryUrl, reloadToken])

  const updateDocumentData = useCallback(
    (newDocData: Partial<DocumentData>) => {
      const merged = deepMerge(documentData, newDocData)
      if (JSON.stringify(documentData) !== JSON.stringify(merged)) {
        setDocumentData(merged)
      }
    },
    [documentData],
  )

  if (!documentData.idOrPath) {
    return null
  }

  return (
    <DocumentDataContext.Provider value={{ documentData, updateDocumentData, isInProgress, triggerReload }}>
      {children}
    </DocumentDataContext.Provider>
  )
}
