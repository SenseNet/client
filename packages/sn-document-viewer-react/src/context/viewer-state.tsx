import { deepMerge, DeepPartial, ObservableValue } from '@sensenet/client-utils'
import React, { useCallback, useEffect, useState } from 'react'
import { ViewerState } from '../models/viewer-state'

export const defaultViewerState: ViewerState & {
  updateState: (newState: DeepPartial<ViewerState>) => void
} = {
  activePage: 1,
  zoomMode: 'custom',
  showWatermark: false,
  showRedaction: true,
  showShapes: true,
  showThumbnails: false,
  fitRelativeZoomLevel: 0,
  rotation: undefined,
  showComments: false,
  hasChanges: false,
  isPlacingCommentMarker: false,
  isCreateCommentActive: false,
  onPageChange: new ObservableValue(1),
  updateState: () => {},
}
export const ViewerStateContext = React.createContext(defaultViewerState)

export const ViewerStateProvider: React.FC<{ options?: Partial<typeof defaultViewerState> }> = (props) => {
  const [state, setState] = useState<typeof defaultViewerState>(deepMerge({ ...defaultViewerState }, props.options))

  useEffect(() => {
    setState(deepMerge({ ...defaultViewerState }, props.options))
  }, [props.options])

  const updateState = useCallback(
    (newState: DeepPartial<typeof defaultViewerState>) => {
      setState(deepMerge({ ...state }, newState))
    },
    [state],
  )

  return <ViewerStateContext.Provider value={{ ...state, updateState }}>{props.children}</ViewerStateContext.Provider>
}
