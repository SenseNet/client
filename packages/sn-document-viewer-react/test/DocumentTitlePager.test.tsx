import React from 'react'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import { act } from 'react-dom/test-utils'
import { mount, shallow } from 'enzyme'
import { DocumentTitlePager } from '../src/components/document-widgets/DocumentTitlePager'
import { defaultViewerState, ViewerStateContext } from '../src/context/viewer-state'
import { DocumentDataContext } from '../src/context/document-data'
import { exampleDocumentData } from './__Mocks__/viewercontext'

describe('DocumentTitlePager component', () => {
  it('Should render the unfocused state without crashing', () => {
    const wrapper = shallow(<DocumentTitlePager />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Should render the focused state without crashing', async () => {
    const wrapper = shallow(<DocumentTitlePager />)
    wrapper.find(Typography).simulate('click')
    expect(wrapper.find(TextField)).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('Should set the default value of the textfield when changed', async () => {
    const updateState = jest.fn()
    const wrapper = mount(
      <ViewerStateContext.Provider
        value={{
          ...defaultViewerState,
          updateState,
        }}>
        <DocumentDataContext.Provider
          value={{
            documentData: {
              ...exampleDocumentData,
              pageCount: 100,
            },
            updateDocumentData: async () => undefined,
          }}>
          <DocumentTitlePager />
        </DocumentDataContext.Provider>
      </ViewerStateContext.Provider>,
    )
    act(() => {
      wrapper.find(Typography).simulate('click')
    })
    act(() => {
      const onChange = wrapper.update().find(TextField).prop('onChange')
      onChange && onChange({ currentTarget: { value: '3' } } as any)
    })
    expect(updateState).toBeCalledWith({ activePages: [3] })
  })
})
