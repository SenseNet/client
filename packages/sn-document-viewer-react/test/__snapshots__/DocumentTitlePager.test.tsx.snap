// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentTitlePager component Should render the focused state without crashing 1`] = `
<ClickAwayListener
  onClickAway={[Function]}
>
  <WithStyles(ForwardRef(Typography))
    color="inherit"
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "cursor": "pointer",
        "display": "flex",
        "margin": "0 2.5em",
        "overflow": "hidden",
      }
    }
    title=""
    variant="h6"
  >
    <div
      style={
        Object {
          "overflow": "hidden",
          "textOverflow": "ellipsis",
          "whiteSpace": "nowrap",
        }
      }
    >
       
    </div>
    <form
      onSubmit={[Function]}
    >
      <WithStyles(ForwardRef(TextField))
        InputLabelProps={
          Object {
            "shrink": true,
          }
        }
        inputProps={
          Object {
            "max": -1,
            "min": 1,
            "style": Object {
              "textAlign": "center",
            },
          }
        }
        margin="dense"
        onChange={[Function]}
        required={true}
        style={
          Object {
            "flexShrink": 0,
          }
        }
        title="Goto page"
        type="number"
        value={1}
      />
    </form>
  </WithStyles(ForwardRef(Typography))>
</ClickAwayListener>
`;

exports[`DocumentTitlePager component Should render the unfocused state without crashing 1`] = `
<ClickAwayListener
  onClickAway={[Function]}
>
  <WithStyles(ForwardRef(Typography))
    color="inherit"
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "cursor": "pointer",
        "display": "flex",
        "margin": "0 2.5em",
        "overflow": "hidden",
      }
    }
    title=""
    variant="h6"
  >
    <div
      style={
        Object {
          "overflow": "hidden",
          "textOverflow": "ellipsis",
          "whiteSpace": "nowrap",
        }
      }
    >
       
    </div>
    <div
      style={
        Object {
          "flexShrink": 0,
        }
      }
    >
      1
       / 
      -1
    </div>
  </WithStyles(ForwardRef(Typography))>
</ClickAwayListener>
`;
